### PR TITLE
Add dtype to ToCupy

### PR DESCRIPTION
**Pull request by drbeh**
### Description
This PR allow `ToCupy` transform define the expected output `dtype`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

Thanks for your enhancement.
May I know your use case for this `ToCupy` transform? I think `cuCIM` already outputs data in cupy array?

Thanks.

**Comment by drbeh** [comment]
> Hi @drbeh ,
> 
> Thanks for your enhancement.
> May I know your use case for this `ToCupy` transform? I think `cuCIM` already outputs data in cupy array?
> 
> Thanks.

Hi @Nic-Ma, you're right! cuCIM output data in the cupy array but it also expect the input data to be in `cupy.ndarray` and in specific data type (if needed). We already had converting to `cupy.ndarray` and this PR add the ability to define `dtype` for that array.

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

I see, could you please help provide the program of `cuCIM` loading part? I am a little bit confused by `it also expects the input data to be in cupy.ndarray`, I thought `cuCIM` reads data from file names.

Thanks.

**Comment by drbeh** [comment]
> Hi @drbeh ,
> 
> I see, could you please help provide the program of `cuCIM` loading part? I am a little bit confused by `it also expects the input data to be in cupy.ndarray`, I thought `cuCIM` reads data from file names.
> 
> Thanks.

That's a good point. I see where the confusion come from. cuCIM team has developed several transforms that we are going to use them in MONAI in conjunction with other transforms (for instance the torch.Tensor based transforms) via a `CuCIM` wrapper (https://github.com/Project-MONAI/MONAI/pull/2932). Thus, although cucim has data loading capabilities, a user can use transforms independently.

**Comment by Nic-Ma** [comment]
OK, I see, thanks for your explanation.
Maybe you can use `CastToType` transform before `ToCupy` transform to achieve expected dtype?
I think it's better to make the `ToCupy` API align with `ToTensor` and `ToNumpy`, etc.
@wyli @rijobro What do you guys think?

Thanks in advance.

**Comment by drbeh** [comment]
> OK, I see, thanks for your explanation.
> Maybe you can use `CastToType` transform before `ToCupy` transform to achieve expected dtype?
> I think it's better to make the `ToCupy` API align with `ToTensor` and `ToNumpy`, etc.
> @wyli @rijobro What do you guys think?
> 
> Thanks in advance.

Thank you @Nic-Ma for your suggestion. What you are saying would work but will create a redundant copy of the image.
For instance if we have a `numpy.ndarray` of `np.uint8` and use `ToCupy`, it creates a `cupy.ndarray` with dtype `np.uint8`. Then if we call `CastToType(np.float32)`, it will create another  `cupy.ndarray` with dtype `np.float32`. 

This is not efficient specially for large images.I believe we should change the `ToTensor` and `ToNumpy` to include dtype too.

**Comment by drbeh** [comment]
@Nic-Ma @wyli , I'd really appreciate your help in reviewing the cucim related PRs since we are depending on them for cucim integration and digital pathology performance optimization, which is supposed to be done by the end of the week,

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

OK, I see your point, as I said in https://github.com/Project-MONAI/MONAI/pull/2926#discussion_r706524927, to avoid to add more and more args, I will submit a PR to support `kwargs` directly soon.

Thanks.

**Comment by Nic-Ma** [comment]
Submitted ticket to track it: https://github.com/Project-MONAI/MONAI/issues/2951
Thanks for raising this topic.

**Comment by drbeh** [comment]
> Submitted ticket to track it: #2951
> Thanks for raising this topic.

@Nic-Ma specifically for `ToCupy` it only supports `dtype` and does not support `device`, so don't you think that can we have this one without kwargs and only with `dtype`?

File: monai/transforms/utility/dictionary.py [hunk 1]
```diff
@@ -553,15 +553,15 @@ class ToCupyd(MapTransform):
 
     backend = ToCupy.backend
 
-    def __init__(self, keys: KeysCollection, allow_missing_keys: bool = False) -> None:
+    def __init__(self, keys: KeysCollection, allow_missing_keys: bool = False, dtype=None) -> None:
         """
         Args:
             keys: keys of the corresponding items to be transformed.
                 See also: :py:class:`monai.transforms.compose.MapTransform`
```
Comment (Nic-Ma) [inline monai/transforms/utility/dictionary.py:560]: Could you please help add the doc-string for `dtype` arg?

Thanks.

File: monai/utils/type_conversion.py [hunk 2]
```diff
@@ -154,6 +155,49 @@ def convert_to_numpy(data, wrap_sequence: bool = False):
...
+            Tensor, numpy array, cupy array, float, int, bool are converted to cupy arrays
+
+            for dictionary, list or tuple, convert every item to a numpy array if applicable.
+        wrap_sequence: if `False`, then lists will recursively call this function. E.g., `[1, 2]` -> `[array(1), array(2)]`.
+            If `True`, then `[1, 2]` -> `array([1, 2])`.
+    """
+
+    # direct calls
+    if isinstance(data, cp_ndarray):
+        if dtype is not None:
```
Comment (Nic-Ma) [inline monai/utils/type_conversion.py:174]: I think you don't need to check `None`.

Thanks.

File: monai/utils/type_conversion.py [hunk 3]
```diff
@@ -154,6 +155,49 @@ def convert_to_numpy(data, wrap_sequence: bool = False):
...
+            data = data.astype(dtype)
+    elif isinstance(data, np.ndarray):
+        data = cp.asarray(data, dtype)
+    elif isinstance(data, torch.Tensor):
+        data = cp.asarray(data, dtype)
+    elif isinstance(data, (float, int, bool)):
+        data = cp.asarray(data, dtype)
+    # recursive calls
+    elif isinstance(data, Sequence) and wrap_sequence:
+        return cp.asarray(data)
```
Comment (Nic-Ma) [inline monai/utils/type_conversion.py:184]: Seems all the above cases can simplified by `cp.asarray(data, dtype)`.

Thanks.

File: monai/utils/type_conversion.py [hunk 4]
```diff
@@ -199,6 +245,11 @@ def convert_data_type(
             data = convert_to_numpy(data)
         if data is not None and dtype != data.dtype:
             data = data.astype(dtype)
+    elif has_cp and output_type is cp.ndarray:
+        if orig_type is not cp.ndarray:
```
Comment (Nic-Ma) [inline monai/utils/type_conversion.py:249]: Seems the `convert_to_cupy` API can support `cupy arrary` input, so you don't need to check it here again.
Just:
```py
elif has_cp and output_type is cp.ndarray:
    if data is not None:
        data = convert_to_cupy(data, dtype)
```
Right?

File: tests/test_to_cupy.py [hunk 5]
```diff
@@ -22,49 +22,81 @@
 cp, has_cp = optional_import("cupy")
 
 
+@skipUnless(has_cp, "CuPy is required.")
 class TestToCupy(unittest.TestCase):
-    @skipUnless(has_cp, "CuPy is required.")
     def test_cupy_input(self):
-        test_data = cp.array([[1, 2], [3, 4]])
+        test_data = cp.array([[1, 2], [3, 4]], dtype=cp.float32)
```
Comment (Nic-Ma) [inline tests/test_to_cupy.py:28]: I am not familiar with Cupy APIs, may I know how to specify the target `device` for the array?

Thanks.

File: monai/utils/type_conversion.py (hunk 3 above)
Comment (drbeh) [reply monai/utils/type_conversion.py:184]: that's a good point. thanks

File: monai/utils/type_conversion.py (hunk 2 above)
Comment (drbeh) [reply monai/utils/type_conversion.py:174]: In cupy, `astype(None)` is equivalnet of `astype('float64')`, so I don't want to change the dtype if it is not intended to.

File: monai/utils/type_conversion.py (hunk 4 above)
Comment (drbeh) [reply monai/utils/type_conversion.py:249]: Good observation but similar to numpy, in cupy, `astype` will make a copy by default so I don't want to `astype` being called if the type and dtype of input and output are the same. https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html#numpy.ndarray.astype

File: monai/utils/type_conversion.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/utils/type_conversion.py:174]: Hi @drbeh ,

I checked the doc-string of `cupy.asarray()`, seems `None` means "It is inferred from the input by default":
https://docs.cupy.dev/en/stable/reference/generated/cupy.asarray.html

Thanks.

File: monai/utils/type_conversion.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/utils/type_conversion.py:249]: I think maybe you mixed `asarray()` and `astype()` here? You can see the doc-string of `asarray()`: "If a is already on the device, no copy is performed."
https://docs.cupy.dev/en/stable/reference/generated/cupy.asarray.html

Thanks.

File: tests/test_to_cupy.py (hunk 5 above)
Comment (drbeh) [reply tests/test_to_cupy.py:28]: CuPy arrays are  always on GPU! You can change the cuda device with `cp.cuda.Device(1).use()` or with a context manager!
Basically CuPy array on CPU is a NumPy array!

Comment (Nic-Ma) [reply tests/test_to_cupy.py:28]: `cp.cuda.Device(1).use()` is what I asked, I know Cupy array is on GPU, I just want to know how to specify GPU device in multi-GPUs case.

Thanks.

File: monai/utils/type_conversion.py (hunk 4 above)
Comment (drbeh) [reply monai/utils/type_conversion.py:249]: Right, I have `astype` in the code but I can change it to `asarray` thanks.

File: tests/test_to_cupy.py (hunk 5 above)
Comment (drbeh) [reply tests/test_to_cupy.py:28]: I see. I like your forward thinking! ;)

**Comment by drbeh** [comment]
Hi @Nic-Ma, thank you very much for your constructive comment. I have applied all of them. Please let me know if there is anything else. thanks

File: monai/utils/type_conversion.py [hunk 6]
```diff
@@ -154,6 +155,42 @@ def convert_to_numpy(data, wrap_sequence: bool = False):
...
+        wrap_sequence: if `False`, then lists will recursively call this function. E.g., `[1, 2]` -> `[array(1), array(2)]`.
+            If `True`, then `[1, 2]` -> `array([1, 2])`.
+    """
+
+    # direct calls
+    if isinstance(data, (cp_ndarray, np.ndarray, torch.Tensor, float, int, bool)):
+        data = cp.asarray(data, dtype)
+    # recursive calls
+    elif isinstance(data, Sequence) and wrap_sequence:
+        return cp.asarray(data)
```
Comment (Nic-Ma) [inline monai/utils/type_conversion.py:177]: Maybe you should also specify `dtype` for a sequence array.

**Review by Nic-Ma — APPROVED** [review]

**Pull request merged**